DeepStorage/Cache/StackableThingComparer.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Verse;
+
+namespace LWM.DeepStorage
+{
+    public class StackableThing_Comparer : IEqualityComparer<Thing>
+    {
+        public static StackableThing_Comparer Instance = new StackableThing_Comparer();
+
+        private StackableThing_Comparer()
+        {
+        }
+
+        #region Implementation of IEqualityComparer<in Thing>
+
+        public bool Equals(Thing x, Thing y)
+        {
+            if (x == y)
+                return true;
+
+            if (x == null || y == null)
+                return false;
+
+            return x.CanStackWith(y);
+        }
+
+        public int GetHashCode(Thing obj)
+        {
+            return (obj.def.GetHashCode() * 397) ^ (obj.Stuff?.GetHashCode() ?? 0);
```

Comment:
Why the stuff mixed in with the hashcode?

Reply:
It is a custom equality comparer for retrieving item from a hash table.
It is used in created the NonFullThings Dictionary in https://github.com/Mhburg/RimWorld-LWM.DeepStorage/blob/master/DeepStorage/Cache/Deep_Storage_Cell_Storage_Model.cs

The NonFullThings dictionary is used to keep track of things which don't have a full stack.

Say, we have two stacks of meat, stack A has a stackCount of 60, stack B has 50. Stack A has been added to the NonFullThings, and now a pawn tried to put down stack B. To find stack A in NonFullThings, the code can query NonFullThings this way NonFullThings[StackB} instead of going through a list.

I will add more comments to my PR when I get around to it.

Reply:
Vanilla also checks if the comps allow stacking?

Reply:
(if it's a ThingWithComps)

Reply:
There are two overrides to this method, one is ThingWithComps and the other is MinifiedThing. They both run base.CanStackWith() first, then run their own logic.